language/move-vm/runtime/src/runtime.rs

```diff
@@ -201,24 +204,54 @@ impl<'alloc> VMRuntime<'alloc> {
     }
 }
 
+fn verify_type_args(signature: &FunctionSignature, ty_args: &[Type]) -> VMResult<()> {
+    if signature.type_formals.len() != ty_args.len() {
+        return Err(
+            VMStatus::new(StatusCode::CONTRAINT_KIND_MISMATCH).with_message(format!(
+                "type argument length mismatch: expected {} got {}",
+                signature.type_formals.len(),
+                ty_args.len()
+            )),
+        );
+    }
+    for (k, ty) in signature.type_formals.iter().zip(ty_args.iter()) {
+        match k {
+            Kind::All | Kind::Resource => {
+                return Err(
+                    VMStatus::new(StatusCode::VERIFIER_INVARIANT_VIOLATION).with_message(
+                        "scripts can only have unrestricted type parameters".to_string(),
+                    ),
+                )
+            }
+            Kind::Unrestricted => (),
+        }
+        if ty.is_resource()? {
```

Comment:
Yeah I'll add documentation here to make the intent clearer.

Reply:
I don't understand this check. We need resource instantiation no? The instantiation has nothing to do with the arguments and it may very well be to direct the code to do the right thing towards a resource, no?

Reply:
You are totally right. I messed this part up as Todd pointed out. I've already got it re-implemented on my local branch, will update soon.